hw17 grpc

DmitryLapitsky opened the pull request:
Домашнее задание 17 grpc

DmitryLapitsky committed:
hw17 grpc

DmitryLapitsky committed:
hw17 grpc

petrelevich commented on hw17_grpc/src/main/java/ru/otus/protobuf/client/CurrentValue.java:
```diff
@@ -0,0 +1,36 @@
+
+    AtomicBoolean increase = new AtomicBoolean(true);
+
+    public synchronized void action(int iterations) {
```
зачем на этом методе synchronized?

DmitryLapitsky replied:
убрал, осталось от предыдущего проекта

petrelevich commented on hw17_grpc/src/main/java/ru/otus/protobuf/GRPCClientNumbers.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    static void handleResponce(FromServer value, CurrentValue clientCounter){
+        int currentValue = clientCounter.getCurrentValue();
```
Получается, что значение, которе выводится в лог изменяется при получении нового значения от сервера.
По ДЗ в лог должен выподиться поток значений, к которому добавляется значение от сервера.

DmitryLapitsky replied:
Очень тонкая грань, вроде эта логика и была реализована. Переделал на всякий, как в видео, хотя и с использованием многопоточности, как уже было. Единая формула: value=value+serverValue+1, после которой значение с сервера устанавливается =0, пока заново от сервера что не прилетит.
Но стало короче и лаконичнее, спасибо)

DmitryLapitsky committed:
hw17 grpc
Работа над ошибками: логика сервера переделана, как на видео. Класс-обработчик сервера переименован под более логичное назнвание.
В клиенте логика осталась прежней, кажется, но переделана. Теперь значение от сервера передается в класс расчета клиента.

DmitryLapitsky committed:
hw17 grpc
Работа над ошибками: логика сервера переделана, как на видео. Класс-обработчик сервера переименован под более логичное назнвание.
В клиенте логика осталась прежней, кажется, но переделана. Теперь значение от сервера передается в класс расчета клиента.
убран мусор

petrelevich commented on hw17_grpc/src/main/java/ru/otus/protobuf/client/CurrentValue.java:
```diff
@@ -0,0 +1,28 @@
+
+public class CurrentValue {
+
+    private int serverValue = 0;
```
на этой переменной сейчас гонка.
один поток проставляет значение (в методе onNext), другой это значение читает (в методе action)
Нужна синхронизация.

DmitryLapitsky committed:
hw17 исправлена ошибка синхронизации, добавлен synchronized к методу setServerValue

petrelevich commented on hw17_grpc/src/main/java/ru/otus/protobuf/client/CurrentValue.java:
```diff
@@ -0,0 +1,28 @@
+
+    private final AtomicInteger currentValue = new AtomicInteger(0);
+
+    public synchronized void setServerValue(long serverValue){
```
чтобы было ребро hb, надо захватить монитор из одного потока, из другого отпустить.
сейчас есть только одна часть, из одного потока.
надо добавить вторую - 
synchronized int getServerValue
и читать значение через нее.

petrelevich replied:
Зачем делать cast (int)serverValue; ?
почему бы ни сделать serverValue типа long.

petrelevich commented on hw17_grpc/src/main/java/ru/otus/protobuf/client/CurrentValue.java:
```diff
@@ -0,0 +1,28 @@
+
+    private int serverValue = 0;
+
+    private final AtomicInteger currentValue = new AtomicInteger(0);
```
не понятно, почему currentValue - это Atomic.
ведь эта переменная используется только в одном потоке.
